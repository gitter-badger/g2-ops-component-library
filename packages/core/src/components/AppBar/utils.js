import { compose } from 'ramda'

const FIND_ALL_UNDERSCORES = /_+/g
const FIND_FIRST_LETTER_REGEX = /\w\S*/g

export const replaceUnderscoreWithSpace = (string) =>
  string && string.length > 0 && string.replace(FIND_ALL_UNDERSCORES, ' ')

export const capitalizeFirstLetter = (string) =>
  string &&
  string.replace(
    FIND_FIRST_LETTER_REGEX,
    (match) => match && `${match.charAt(0).toUpperCase()}${match.substr(1).toLowerCase()}`
  )

export const beautifyRoleText = compose(capitalizeFirstLetter, replaceUnderscoreWithSpace)