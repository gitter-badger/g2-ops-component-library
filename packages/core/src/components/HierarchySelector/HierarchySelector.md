Hierarchy Selector Example

Simple Example:
```js
const Example = require('./Example').default
;<Example />
```

Tow Provider Example:
```js
const TowProviderExample = require('./ComplexExample').default
;<TowProviderExample />
```
