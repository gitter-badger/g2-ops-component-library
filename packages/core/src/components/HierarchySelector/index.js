// @flow

export { HierarchySelector } from './HierarchySelector'
