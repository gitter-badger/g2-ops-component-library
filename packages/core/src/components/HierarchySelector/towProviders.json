{
  "status": "success",
  "data": [
    {
      "vendor_id": 226,
      "cas_reference_id": 900000492,
      "country_code": "USA",
      "vendor_name": "G&R TRANSPORTATION",
      "vendor_short_name": "G/RTOW",
      "contact_name": "JUSTIN BARGE",
      "first_name": "",
      "insurance_expiration_date": "2016-07-02T00:00:00-07:00",
      "middle_name": "",
      "last_name": "RICHARD BARGE",
      "email": "obfuscated@copart.com",
      "primary_address_id": null,
      "parent_vendor_id": 0,
      "primary_payment_id": null,
      "vendor_type_code": "DIR",
      "corporate_type_code": "P",
      "legal_id": "",
      "tax_id": "680259420",
      "total_capacity": null,
      "number_of_trucks": 0,
      "daily_capacity": 0,
      "sync_status": "N",
      "dispatch_flag": true,
      "status": "A",
      "created_at": "2018-01-08T11:41:18-08:00",
      "created_by": "DATA MIG",
      "updated_at": "2018-01-08T11:41:18-08:00",
      "updated_by": "DATA MIG",
      "source_system": "CAS",
      "dispatch_groups": [
        {
          "dispatch_group_id": 203,
          "dispatch_group_name": "G/RTOWDG",
          "dispatch_group_description": "G/RTOW",
          "default_flag": "Y",
          "vendor_id": 226,
          "dispatch_flag": true,
          "status": "A",
          "created_at": "2018-01-09T09:13:04-08:00",
          "created_by": "DATA MIG",
          "updated_at": "2018-01-09T09:13:04-08:00",
          "updated_by": "DATA MIG",
          "source_system": "CAS",
          "personnel": [
            {
              "vendor_personnel_id": 1000,
              "vendor_id": 226,
              "role_id": null,
              "p_card_flag": "Y",
              "first_name": "ALLEN",
              "middle_name": null,
              "last_name": "MARQUEZ",
              "date_of_birth": null,
              "is_copart_employee": "N",
              "social_security_number": "999999999",
              "start_date": null,
              "end_date": "2018-02-01T00:00:00-08:00",
              "comments": null,
              "dispatch_flag": true,
              "status": "A",
              "created_at": "2018-01-08T11:48:59-08:00",
              "created_by": "DATA MIG",
              "updated_at": "2018-01-08T11:48:59-08:00",
              "updated_by": "DATA MIG",
              "source_system": "CAS",
              "phone_number": "1112223334",
              "phone_extension": "5555",
              "short_name": null
            },
            {
              "vendor_personnel_id": 2821,
              "vendor_id": 226,
              "role_id": null,
              "p_card_flag": "Y",
              "first_name": "ROBERT",
              "middle_name": null,
              "last_name": "PAQUETTE",
              "date_of_birth": null,
              "is_copart_employee": "N",
              "social_security_number": "999999999",
              "start_date": null,
              "end_date": "2018-02-01T00:00:00-08:00",
              "comments": null,
              "dispatch_flag": true,
              "status": "A",
              "created_at": "2018-01-08T11:48:59-08:00",
              "created_by": "DATA MIG",
              "updated_at": "2018-01-08T11:48:59-08:00",
              "updated_by": "DATA MIG",
              "source_system": "CAS",
              "phone_number": "1112223334",
              "phone_extension": "5555",
              "short_name": null
            },
            {
              "vendor_personnel_id": 1282,
              "vendor_id": 226,
              "role_id": null,
              "p_card_flag": "Y",
              "first_name": "JAMES",
              "middle_name": null,
              "last_name": "PEPPERS",
              "date_of_birth": null,
              "is_copart_employee": "N",
              "social_security_number": "999999999",
              "start_date": null,
              "end_date": "2018-02-01T00:00:00-08:00",
              "comments": null,
              "dispatch_flag": true,
              "status": "A",
              "created_at": "2018-01-08T11:48:59-08:00",
              "created_by": "DATA MIG",
              "updated_at": "2018-01-08T11:48:59-08:00",
              "updated_by": "DATA MIG",
              "source_system": "CAS",
              "phone_number": "1112223334",
              "phone_extension": "5555",
              "short_name": null
            }
          ]
        }
      ]
    },
    {
      "vendor_id": 42511,
      "cas_reference_id": 910042511,
      "country_code": "USA",
      "vendor_name": "PENNY HOFSTADTER",
      "vendor_short_name": "PEHOFSTADT",
      "contact_name": "PENNY HOFSTADTER",
      "first_name": null,
      "middle_name": null,
      "insurance_expiration_date": "2018-07-02T00:00:00-07:00",
      "last_name": null,
      "email": null,
      "primary_address_id": null,
      "parent_vendor_id": null,
      "primary_payment_id": null,
      "vendor_type_code": "INT",
      "corporate_type_code": null,
      "legal_id": null,
      "tax_id": null,
      "total_capacity": null,
      "number_of_trucks": null,
      "daily_capacity": null,
      "sync_status": "",
      "dispatch_flag": true,
      "status": "I",
      "created_at": "2018-03-16T14:03:04-07:00",
      "created_by": "service-account@copart.com",
      "updated_at": "2018-03-16T14:03:04-07:00",
      "updated_by": "service-account@copart.com",
      "source_system": "ops",
      "dispatch_groups": [
        {
          "dispatch_group_id": 40342,
          "dispatch_group_name": "DEFAULT DISPATCH GROUP",
          "dispatch_group_description": "Default Dispatch Group",
          "default_flag": "Y",
          "vendor_id": 42511,
          "dispatch_flag": true,
          "status": "A",
          "created_at": "2018-03-16T14:03:04-07:00",
          "created_by": "service-account@copart.com",
          "updated_at": "2018-03-16T14:03:04-07:00",
          "updated_by": "service-account@copart.com",
          "source_system": "ops",
          "personnel": [
            {
              "vendor_personnel_id": 4543,
              "vendor_id": 42511,
              "role_id": null,
              "p_card_flag": "Y",
              "first_name": "MICHAEL",
              "middle_name": null,
              "last_name": "SPRIFKE",
              "date_of_birth": null,
              "is_copart_employee": "N",
              "social_security_number": "999999999",
              "start_date": null,
              "end_date": "2018-02-01T00:00:00-08:00",
              "comments": null,
              "dispatch_flag": false,
              "status": "A",
              "created_at": "2018-01-08T11:48:59-08:00",
              "created_by": "DATA MIG",
              "updated_at": "2018-01-08T11:48:59-08:00",
              "updated_by": "DATA MIG",
              "source_system": "CAS",
              "phone_number": "1112223334",
              "phone_extension": "5555",
              "short_name": null
            },
            {
              "vendor_personnel_id": 21421,
              "vendor_id": 42511,
              "role_id": null,
              "p_card_flag": "Y",
              "first_name": "MIRANDA",
              "middle_name": null,
              "last_name": "HOLLOWAY",
              "date_of_birth": null,
              "is_copart_employee": "N",
              "social_security_number": "999999999",
              "start_date": null,
              "end_date": "2018-02-01T00:00:00-08:00",
              "comments": null,
              "dispatch_flag": true,
              "status": "A",
              "created_at": "2018-01-08T11:48:59-08:00",
              "created_by": "DATA MIG",
              "updated_at": "2018-01-08T11:48:59-08:00",
              "updated_by": "DATA MIG",
              "source_system": "CAS",
              "phone_number": "1112223334",
              "phone_extension": "5555",
              "short_name": null
            },
            {
              "vendor_personnel_id": 15235,
              "vendor_id": 42511,
              "role_id": null,
              "p_card_flag": "Y",
              "first_name": "DEVIN",
              "middle_name": null,
              "last_name": "BURT",
              "date_of_birth": null,
              "is_copart_employee": "N",
              "social_security_number": "999999999",
              "start_date": null,
              "end_date": "2019-02-01T00:00:00-08:00",
              "comments": null,
              "dispatch_flag": true,
              "status": "A",
              "created_at": "2018-01-08T11:48:59-08:00",
              "created_by": "DATA MIG",
              "updated_at": "2018-01-08T11:48:59-08:00",
              "updated_by": "DATA MIG",
              "source_system": "CAS",
              "phone_number": "1112223334",
              "phone_extension": "5555",
              "short_name": null
            },
            {
              "vendor_personnel_id": 12862,
              "vendor_id": 42511,
              "role_id": null,
              "p_card_flag": "Y",
              "first_name": "ALI",
              "middle_name": null,
              "last_name": "KHALIL",
              "date_of_birth": null,
              "is_copart_employee": "N",
              "social_security_number": "999999999",
              "start_date": null,
              "end_date": "2018-02-01T00:00:00-08:00",
              "comments": null,
              "dispatch_flag": false,
              "status": "A",
              "created_at": "2018-01-08T11:48:59-08:00",
              "created_by": "DATA MIG",
              "updated_at": "2018-01-08T11:48:59-08:00",
              "updated_by": "DATA MIG",
              "source_system": "CAS",
              "phone_number": "1112223334",
              "phone_extension": "5555",
              "short_name": null
            }
          ]
        }
      ]
    },
    {
      "vendor_id": 42517,
      "cas_reference_id": 910042517,
      "country_code": "USA",
      "vendor_name": "LEONARD HOFSTADTER",
      "vendor_short_name": "LEHOFSTADT",
      "contact_name": "LEONARD HOFSTADTER",
      "first_name": null,
      "middle_name": null,
      "last_name": null,
      "email": "leonard@copart-test.com",
      "primary_address_id": null,
      "parent_vendor_id": null,
      "primary_payment_id": null,
      "vendor_type_code": "INT",
      "corporate_type_code": "N",
      "legal_id": null,
      "tax_id": "999999999",
      "total_capacity": null,
      "number_of_trucks": null,
      "daily_capacity": null,
      "sync_status": "",
      "dispatch_flag": true,
      "status": "I",
      "created_at": "2018-03-16T14:07:02-07:00",
      "created_by": "service-account@copart.com",
      "updated_at": "2018-03-16T15:35:12-07:00",
      "updated_by": "Michael.Groat@copart.com",
      "source_system": "ops",
      "dispatch_groups": [
        {
          "dispatch_group_id": 40348,
          "dispatch_group_name": "DEFAULT DISPATCH GROUP",
          "dispatch_group_description": "Default Dispatch Group",
          "default_flag": "Y",
          "vendor_id": 42517,
          "dispatch_flag": true,
          "status": "A",
          "created_at": "2018-03-16T14:07:03-07:00",
          "created_by": "service-account@copart.com",
          "updated_at": "2018-03-16T14:07:03-07:00",
          "updated_by": "service-account@copart.com",
          "source_system": "ops",
          "personnel": [
            {
              "vendor_personnel_id": 15235,
              "vendor_id": 42517,
              "role_id": null,
              "p_card_flag": "Y",
              "first_name": "DEVIN",
              "middle_name": null,
              "last_name": "BURT",
              "date_of_birth": null,
              "is_copart_employee": "N",
              "social_security_number": "999999999",
              "start_date": null,
              "end_date": "2019-02-01T00:00:00-08:00",
              "comments": null,
              "dispatch_flag": true,
              "status": "A",
              "created_at": "2018-01-08T11:48:59-08:00",
              "created_by": "DATA MIG",
              "updated_at": "2018-01-08T11:48:59-08:00",
              "updated_by": "DATA MIG",
              "source_system": "CAS",
              "phone_number": "1112223334",
              "phone_extension": "5555",
              "short_name": null
            }
          ]
        }
      ]
    }
  ]
}
