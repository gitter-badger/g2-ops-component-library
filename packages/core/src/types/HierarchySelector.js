export type OptionType = {
  id: string,
  label: string,
  level: string,
  options: Array<OptionType>,
}

export type FlattenedOptionType = {
  id: string,
  label: string,
  level: string,
  display: any,
  path: Array<string>,
  hierarchy: Object,
}
