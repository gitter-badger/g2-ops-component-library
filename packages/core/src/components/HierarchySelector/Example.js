import React from 'react'
import { FlattenedOptionType } from 'types/HierarchySelector'

import { HierarchySelector } from './HierarchySelector'

export const nestedOptions = [
  {
    id: 'asia',
    label: 'Asia',
    level: 'continent',
    options: [
      {
        id: 'india',
        label: 'India',
        level: 'country',
        options: [
          {
            id: 'telangana',
            label: 'Telangana',
            level: 'state',
          },
        ],
      },
      {
        id: 'china',
        label: 'China',
        level: 'country',
        options: [
          {
            id: 'beijing',
            label: 'Beijing',
            level: 'state',
          },
        ],
      },
    ],
  },
  {
    id: 'northAmerica',
    label: 'North America',
    level: 'continent',
    options: [
      {
        id: 'usa',
        label: 'USA',
        level: 'country',
        options: [
          {
            id: 'texas',
            label: 'Texas',
            level: 'state',
          },
          {
            id: 'arkansas',
            label: 'Arkansas',
            level: 'state',
          },
        ],
      },
    ],
  },
  {
    id: 'africa',
    label: 'Africa',
    level: 'continent',
    options: [
      {
        id: 'egypt',
        label: 'Egypt',
        level: 'country',
      },
      {
        id: 'libya',
        label: 'Libya',
        level: 'country',
      },
    ],
  },
]

export const renderMethod = (option) => (
  <div className="hierarchyOption">
    <span>{option.label}</span>
  </div>
)

const initialValue = {
  continent: 'northAmerica',
  country: 'usa',
  state: 'texas'
}

class Example extends React.Component {
  state = {
    value: initialValue,
  }
  onChange = (selectedOption: FlattenedOptionType) => {
    const selectedHierarchyOption = (typeof selectedOption === 'string')? selectedOption : selectedOption.hierarchy
    this.setState({ value: selectedHierarchyOption })
  }
  render () {
    return (
      <div style={{ maxWidth: '400px' }}>
        <HierarchySelector
          name="Simple Hierarchy Example"
          label="Simple Hierarchy Example"
          options={nestedOptions}
          width={200}
          serializeOption={(o) => o.hierarchy}
          optionStyleProps={{ rowHeight: 40, optionsMinHeight: 200, fontSize: '12px' }}
          value={this.state.value}
          renderMethod={renderMethod}
          onChange={this.onChange}
        />
      </div>
    )
  }
}

export default Example
