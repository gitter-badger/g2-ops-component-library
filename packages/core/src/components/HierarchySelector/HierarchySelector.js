// @flow
import type { Node } from 'react'

import type { OptionsType, FlattenedOptionType } from 'types/HierarchySelector'

import React, { PureComponent } from 'react'
import DownArrow from 'material-ui/svg-icons/hardware/keyboard-arrow-down'
import renderIf from 'render-if'

import {AutoSelect} from 'components/AutoSelect'

import { wrapMuiContext } from 'utilities/wrapMuiContext'
import { flattenNestedOptions } from './hierarchySelector.transformer'

import './HierarchySelector.scss'

type HierarchySelectorPropType = {
  options: OptionsType,
  onFocus: (SyntheticKeyboardEvent<HTMLInputElement>) => void,
  value: number,
  renderMethod: Function,
  onChange: Function,
}

type HierarchySelectorStateType = {
  flattenedOptions: Array<FlattenedOptionType>,
  filteredOptions: Array<FlattenedOptionType>,
  selectedValue: string,
}

const DownArrowIcon = wrapMuiContext(DownArrow)

export class HierarchySelector extends PureComponent<HierarchySelectorPropType, HierarchySelectorStateType> {
  autoSelect: Node // eslint-disable-line
  constructor(props: HierarchySelectorPropType) {
    super(props)
    const { options, value: optionId, renderMethod } = props
    const flattenedOptions = flattenNestedOptions(options, renderMethod)
    const renderedPath = this.convertIdToPath(optionId, flattenedOptions)
    this.state = {
      flattenedOptions,
      filteredOptions: flattenedOptions,
      selectedValue: renderedPath,
    }
  }

  componentWillReceiveProps(nextProps: HierarchySelectorPropType) {
    const { onChange } = this.props
    const { options, value: optionId, renderMethod } = nextProps
    if (options !== this.props.options) {
      const flattenedOptions = flattenNestedOptions(options, renderMethod)
      const renderedPath = this.convertIdToPath(optionId, flattenedOptions)
      this.setState((prevState) => ({
        ...prevState,
        flattenedOptions,
        filteredOptions: flattenedOptions,
        selectedValue: renderedPath,
      }))
    }
  }

  isControlled() {
    return this.props.hasOwnProperty('value')
  }

  onChange = (changedOption: FlattenedOptionType) => {
    const { onChange: onChangeMethod } = this.props
    if (typeof changedOption === 'string') {
      this.setState((prevState) => ({
        ...prevState,
        selectedValue: changedOption,
        filteredOptions: prevState.flattenedOptions.filter((option: FlattenedOptionType) =>
          option.label.toLowerCase().startsWith(changedOption.toLowerCase()),
        ),
      }))
    } else if (!changedOption.isDisabled) {
      const filteredOptions = this.state.flattenedOptions.filter((option: FlattenedOptionType) =>
        option.label.toLowerCase().startsWith(changedOption.label.toLowerCase()),
      )
      this.setState((prevState) => ({
        ...prevState,
        selectedValue: this.renderSelectedOption(changedOption),
        filteredOptions: filteredOptions,
      }))
      // invokes parent's onChange method
      onChangeMethod && onChangeMethod(changedOption)
    }
  }

  onFocus = (e: SyntheticKeyboardEvent<HTMLInputElement>) => {
    this.autoSelect.select()
    if (this.props.onFocus) {
      this.props.onFocus(e)
    }
  }

  convertIdToPath = (lookupId: number, options: Array<FlattenedOptionType>) => {
    const valueAmongOptions = options.find((o) => o.id === lookupId)
    const displayMethod = this.renderSelectedOption || this.renderOption
    const result = valueAmongOptions ? displayMethod(valueAmongOptions) : ''
    return result
  }

  renderOption = (option: FlattenedOptionType) => {
    const numberOfSpaces = option.path.length
    const renderIfHasChildren = renderIf(option.hasChildren)
    const { filteredOptions, flattenedOptions } = this.state
    const isFilteringDone = filteredOptions.length < flattenedOptions.length
    const renderIfFilteringIsDone = renderIf(isFilteringDone)
    const paddingLeft = isFilteringDone ? '0px' : `${(numberOfSpaces - 1) * 25}px`
    const displayTitle = option.path.slice(0, numberOfSpaces - 1).join(' - ')
    return (
      <div style={{ display: 'flex', flexDirection: 'row', paddingLeft }}>
        <div style={{ marginTop: '-5px', display: 'flex', alignItems: 'center' }}>
          {renderIfHasChildren(<i className="material-icons">arrow_drop_down</i>)}
        </div>
        <div style={{ display: 'flex', flexDirection: 'row' }}>
          {renderIfFilteringIsDone(<div className="filtered">{displayTitle}</div>)}
          <div title={`${displayTitle} - ${option.label}`}>{option.display}</div>
        </div>
      </div>
    )
  }

  renderSelectedOption = (option: FlattenedOptionType) => option.path.join(' - ')

  render() {
    const { filteredOptions, selectedValue } = this.state

    return (
      <div className="HierarchySelector">
        <AutoSelect
          ref={(n) => {
            this.autoSelect = n
          }}
          {...this.props}
          options={filteredOptions}
          serializeOption={(o) => o.id}
          displayOption={this.renderOption}
          displaySelectedOption={this.renderSelectedOption}
          onChange={this.onChange}
          value={selectedValue}
          onFocus={this.onFocus}
        />
      </div>
    )
  }
}
