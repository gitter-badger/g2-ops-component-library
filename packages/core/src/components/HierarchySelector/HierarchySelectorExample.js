import React from 'react'
import renderIf from 'render-if'
import { prop } from 'ramda'
import { formatDate, daysElapsedUntilToday } from '../DatePicker/dateUtils'
import { FlattenedOptionType, OptionType } from 'types/HierarchySelector'
import {HierarchySelector} from './HierarchySelector'
import towProvidersJson from './towProviders.json'

// LEHOFSTADT - DEFAULT DISPATCH - GROUP - DEVIN BURT
const initialUser = {
  company: 42517,
  group: 40348,
  driver: 15235,
}

const initialOptionId = prop('driver', initialUser) || prop('group', initialUser) || prop('company', initialUser)

export const CardIcon = () => (
  <div className="CardIcon Icon" data-tip="PCard">
    <div className="red circle" />
    <div className="orange circle" />
  </div>
)

export const renderMethod = (option) => {
  const renderIfCompany = renderIf(option.level === 'company' && option.expiresOn)
  const labelColor = option.dispatch_flag === true ? 'black' : 'grey'
  const expiryDateColor = option.isExpired ? 'red' : 'black'
  const renderIfCard = renderIf(option.p_card_flag === 'Y')
  return (
    <div className="hierarchyOption">
      <span style={{ color: `${labelColor}` }}>{option.label}</span>
      {renderIfCompany(
        <span className="separator">
          <i>
            Expires: <span style={{ color: `${expiryDateColor}` }}>{option.expiresOn} </span>
          </i>
        </span>,
      )}
      {renderIfCard(
        <span className="separator">
          <CardIcon />
        </span>,
      )}
    </div>
  )
}
const isExceeded = (givenDate: string, numberOfDays: number) => daysElapsedUntilToday(givenDate) > numberOfDays

function transformPersonnelOption(option, isVendorExpired) {
  const componentOption: OptionType = {
    id: option.vendor_personnel_id,
    level: 'driver',
    label: `${option.first_name} ${option.last_name}`,
    dispatch_flag: option.dispatch_flag,
    p_card_flag: option.p_card_flag,
    isExpired: isVendorExpired,
    isSelectable: !isVendorExpired && option.dispatch_flag === true,
  }
  return componentOption
}

function transformGroupOption(option, isVendorExpired) {
  const componentOption: OptionType = {
    id: option.dispatch_group_id,
    level: 'group',
    label: option.dispatch_group_name,
    dispatch_flag: option.dispatch_flag,
    p_card_flag: option.p_card_flag,
    isExpired: isVendorExpired,
    isSelectable: !isVendorExpired && option.dispatch_flag === true,
    options: option.personnel.map((personnelOption) => transformPersonnelOption(personnelOption, isVendorExpired)),
  }
  return componentOption
}

function transformVendorOption(option) {
  const formattedDate: OptionType =
    option.insurance_expiration_date && formatDate(option.insurance_expiration_date, 'YYYY-MM-DDTHH:mm:sss Z')
  const isVendorExpired = formattedDate && isExceeded(formattedDate, 30)
  const componentOption = {
    id: option.vendor_id,
    level: 'company',
    label: option.vendor_short_name,
    dispatch_flag: option.dispatch_flag,
    expiresOn: formattedDate,
    p_card_flag: option.p_card_flag,
    isExpired: isVendorExpired,
    isSelectable: !isVendorExpired && option.dispatch_flag === true,
    options: option.dispatch_groups.map((groupOption) => transformGroupOption(groupOption, isVendorExpired)),
  }
  return componentOption
}
const towProviders = towProvidersJson.data
const componentOptions = towProviders.map(transformVendorOption)

class Example extends React.Component {
  state = {
    value: null,
  }
  onChange = (selectedOption: FlattenedOptionType) => {
    const selectedHierarchyOption = (typeof selectedOption === 'string')? selectedOption : selectedOption.hierarchy
    this.setState({ value: selectedHierarchyOption })
  }
  render () {
    return (
      <div style={{ maxWidth: '400px' }}>
        <HierarchySelector
          name="Tow Provider"
          label="Tow Provider"
          options={componentOptions}
          width={200}
          optionStyleProps={{ rowHeight: 40, optionsMinHeight: 200, fontSize: '12px' }}
          value={initialUser}
          renderMethod={renderMethod}
          onChange={this.onChange}
          onRenderSuffix={() => <i className="material-icons">arrow_drop_down</i>}
        />
      </div>
    )
  }
}

export default Example
