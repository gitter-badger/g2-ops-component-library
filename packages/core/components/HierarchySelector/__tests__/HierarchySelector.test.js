import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import renderIf from 'render-if'

import { wrapWithMaterialUIContext } from 'utils/wrapWithContext'

import HierarchySelector from '../HierarchySelector'
import { CardIcon, renderMethod } from '../HierarchySelectorExample'

const nestedOptions = [
  {
    name: 'asia',
    label: 'Asia',
    options: [
      {
        name: 'india',
        label: 'India',
        options: [
          {
            name: 'telangana',
            label: 'Telangana',
          },
        ],
      },
      {
        name: 'china',
        label: 'China',
        options: [
          {
            name: 'beijing',
            label: 'Beijing',
          },
        ],
      },
    ],
  },
  {
    name: 'northAmerica',
    label: 'North America',
    options: [
      {
        name: 'usa',
        label: 'USA',
        options: [
          {
            name: 'texas',
            label: 'Texas',
          },
          {
            name: 'arkansas',
            label: 'Arkansas',
          },
        ],
      },
    ],
  },
  {
    name: 'africa',
    label: 'Africa',
    options: [
      {
        name: 'egypt',
        label: 'Egypt',
      },
      {
        name: 'libya',
        label: 'Libya',
      },
    ],
  },
]

const initialValue = "Haul O Way - Haul O' Way - Jason"

describe('<HierarchySelector />', () => {
  test('should render nested options properly', () => {
    const tree = renderer.create(
      <HierarchySelector
        name="AutoSelect Field"
        options={nestedOptions}
        width={200}
        optionStyleProps={{ rowHeight: 40, optionsMinHeight: 200 }}
        value={initialValue}
        renderMethod={renderMethod}
      />,
    )
    expect(tree).toMatchSnapshot()
  })

  test('should filter options available based on entered text', () => {
    const tree = mount(
      <HierarchySelector
        name="AutoSelect Field"
        options={nestedOptions}
        width={200}
        optionStyleProps={{ rowHeight: 40, optionsMinHeight: 200 }}
        value={initialValue}
        renderMethod={renderMethod}
      />,
    )
    expect(tree.state('filteredOptions')).toEqual(tree.state('flattenedOptions'))
    tree.find('AutoSelect').prop('onChange')('Te')
    expect(tree.state('filteredOptions')).toMatchSnapshot()
  })

  test('should save selected value when value is selected from dropdown', () => {
    const tree = mount(
      <HierarchySelector
        name="AutoSelect Field"
        options={nestedOptions}
        width={200}
        optionStyleProps={{ rowHeight: 40, optionsMinHeight: 200 }}
        value={initialValue}
        renderMethod={renderMethod}
      />,
    )
    tree.find('AutoSelect').prop('onChange')('Te')
    tree.find('AutoSelect').prop('onChange')(tree.state('filteredOptions')[0])
    expect(tree.state('selectedValue')).toMatchSnapshot()
  })
})
