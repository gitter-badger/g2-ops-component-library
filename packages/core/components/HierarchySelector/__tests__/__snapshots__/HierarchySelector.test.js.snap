// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HierarchySelector /> should filter options available based on entered text 1`] = `
Array [
  Object {
    "display": <div
      className="hierarchyOption"
    >
      <span
        style={
          Object {
            "color": "grey",
          }
        }
      >
        Telangana
      </span>
    </div>,
    "hasChildren": false,
    "hierarchy": Object {
      "undefined": undefined,
    },
    "label": "Telangana",
    "name": "telangana",
    "path": Array [
      "Asia",
      "India",
      "Telangana",
    ],
  },
  Object {
    "display": <div
      className="hierarchyOption"
    >
      <span
        style={
          Object {
            "color": "grey",
          }
        }
      >
        Texas
      </span>
    </div>,
    "hasChildren": false,
    "hierarchy": Object {
      "undefined": undefined,
    },
    "label": "Texas",
    "name": "texas",
    "path": Array [
      "North America",
      "USA",
      "Texas",
    ],
  },
]
`;

exports[`<HierarchySelector /> should render nested options properly 1`] = `
<div
  className="HierarchySelector"
>
  <div>
    <div
      className="ms-Fabric css-32"
    >
      <div
        className="ms-TextField root"
      >
        <div
          className="ms-TextField-wrapper wrapper"
        >
          <div
            className="ms-TextField-fieldGroup fieldGroup"
          >
            <input
              aria-describedby={null}
              aria-invalid={false}
              aria-label={undefined}
              autoComplete="off"
              className="ms-TextField-field field input "
              disabled={undefined}
              id="TextField0"
              name="AutoSelect Field-txtField"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              placeholder={undefined}
              title=""
              type="text"
              value=""
              width={200}
            />
          </div>
        </div>
      </div>
    </div>
    <input
      data-uname={undefined}
      name="AutoSelect Field"
      type="hidden"
      value=""
    />
  </div>
</div>
`;

exports[`<HierarchySelector /> should save selected value when value is selected from dropdown 1`] = `"Asia - India - Telangana"`;
